### DISCUSSION TITLE
Where is the file implementing the logic to detect duplication in MATLAB fliles?

### DISCUSSION THREAD

**Discussion by Masfouf** [pmd/pmd#3732] [Q&A] [answered]
Hello,
I would like to know the logic you are using to detect duplication in MATLAB files. 

Please can you explain to me this logic and where is the file implementing it?

Thanks in advance!

**Comment by jsotuyod** [comment, marked as answer]
CPD works the same way for all languages:
1. Based on a language-specific parser, we tokenize each file to be analyzed.
2. We roll over the code and compute a hash on a rolling *n token window* for each file's codebase. [This code](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/MatchAlgorithm.java#L98-L100)
3. We group together token chains that have the same hash. [This code](https://github.com/pmd/pmd/blob/4ae7e40bad8a1c1a9bfada21c3608b80e1c616a8/pmd-core/src/main/java/net/sourceforge/pmd/cpd/MatchAlgorithm.java#L103-L117)

Your main concern is really as to how the code is tokenized, and how strictly you retain the code; for instance, are you looking for exact copy-pastes (exact variable and function names), or not. Some languages (specially Java), support several flags to control these, but not all languages do this to the same extent.

Matlab itself support no additional flag, only has the built-in support to ignore comments and snippets of code within `CPD-OFF` / `CPD-ON` comments. [See the docs](https://pmd.github.io/pmd-6.41.0/pmd_userdocs_cpd.html#suppression)

**Reply by adangel** [reply]
Tokenization of MATLAB files is done by javacc with [this grammer file](https://github.com/pmd/pmd/blob/master/pmd-matlab/etc/grammar/matlab.jj).
